akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ShardingMessageExtractor.scala

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (C) 2017-2018 Lightbend Inc. <http://www.lightbend.com/>
+ */
+package akka.cluster.sharding.typed
+
+object ShardingMessageExtractor {
+
+  /**
+   * Scala API:
+   *
+   * Create the default message extractor, using envelopes to identify what entity a message is for
+   * and the hashcode of the entityId to decide which shard an entity belongs to.
+   *
+   * This is recommended since it does not force details about sharding into the entity protocol
+   */
+  def apply[A](maxNumberOfShards: Int, handOffStopMessage: A): ShardingMessageExtractor[ShardingEnvelope[A], A] =
+    new HashCodeMessageExtractor[A](maxNumberOfShards, handOffStopMessage)
+
+  /**
+   * Scala API: Create a message extractor for a protocol where the entity id is available in each message.
+   */
+  def noEnvelope[A](
+    maxNumberOfShards:  Int,
+    handOffStopMessage: A)(
+    extractEntityId: A ⇒ String): ShardingMessageExtractor[A, A] =
+    new HashCodeNoEnvelopeMessageExtractor[A](maxNumberOfShards, handOffStopMessage) {
+      def entityId(message: A) = extractEntityId(message)
+    }
+
+}
+
+/**
+ * Entirely customizable typed message extractor. Prefer [[HashCodeMessageExtractor]] or
+ * [[HashCodeNoEnvelopeMessageExtractor]]if possible.
+ *
+ * @tparam E Possibly an Envelope around the messages accepted by the entity actor, is the same as `A` if there is no
+ *           envelope.
+ * @tparam A The type of message accepted by the entity actor
+ */
+abstract class ShardingMessageExtractor[E, A] {
+
+  /**
+   * Extract the entity id from an incoming `message`. If `null` is returned
+   * the message will be `unhandled`, i.e. posted as `Unhandled` messages on the event stream
+   */
+  def entityId(message: E): String
+
+  /**
+   * The shard identifier for a given entity id. Only messages that passed the [[ShardingMessageExtractor#entityId]]
+   * function will be used as input to this function.
+   */
+  def shardId(entityId: String): String
+
+  /**
+   * Extract the message to send to the entity from an incoming `message`.
+   * Note that the extracted message does not have to be the same as the incoming
+   * message to support wrapping in message envelope that is unwrapped before
+   * sending to the entity actor.
+   */
+  def unwrapMessage(message: E): A
+
+  /**
+   * Message sent to an entity to tell it to stop, e.g. when rebalanced.
+   * The message defined here is not passed to `entityId`, `shardId` or `unwrapMessage`.
+   */
+  def handOffStopMessage: A
```

Comment:
Do we need to document if this is called once or multiple times?

Reply:
it's currently only called once at startup, because untyped sharding needs it then, but I think we should leave that undefined. A later implementation might call it on demand, and several times. I don't think it matters

Reply:
Leaving undefined is fine, just wanted to know we're doing so consciously :)